flink-clients/src/main/java/org/apache/flink/client/program/Client.java

```diff
@@ -65,31 +73,25 @@
 /**
  * Encapsulates the functionality necessary to submit a program to a remote cluster.
  */
-public class Client {
+public abstract class Client {
```

Comment:
Whoever introduced the savepoint method I think did it out of API compatibility reasons. 

I was wondering the same for the many run methods (actually tried to remove some). We need to duplicate them because runBlocking returns a JobExuectionResult and runDetached returns a JobSubmission result. We could get rid of the duplication if we did some naughty casting depending on the result returned.

Optimized plans are only created for batch. The problem is that there are really that many ways to submit Flink programs. I think we could tackle this in a follow-up PR.